### PR TITLE
Add additional technical analysis algorithms

### PR DESCRIPTION
**Pull request by jsternberg**
Adapted from #9260.

### CODE REVIEW

File: query/call_iterator.go [hunk 1]
```diff
@@ -1247,6 +1247,32 @@ func newMovingAverageIterator(input Iterator, n int, opt IteratorOptions) (Itera
 	}
 }
 
+// newExponentialMovingAverageIterator returns an iterator for operating on an exponential_moving_average() call.
+func newExponentialMovingAverageIterator(input Iterator, n int, nHold int, warmupType string, opt IteratorOptions) (Iterator, error) {
```
Comment (jsternberg) [inline query/call_iterator.go]: `n, nHold int` you don't need to repeat the argument twice.

The argument for `warmupType` should be the `gota.WarmupType`. The string to warmup type can be somewhere else.

File: query/compile.go [hunk 2]
```diff
@@ -550,6 +552,61 @@ func (c *compiledField) compileMovingAverage(args []influxql.Expr) error {
...
+		case *influxql.IntegerLiteral:
+			if arg2.Val < 0 && arg2.Val != -1 {
+				return fmt.Errorf("exponential_moving_average hold period must be greater than or equal to 0")
+			}
+		default:
+			return fmt.Errorf("exponential_moving_average hold period must be an integer")
+		}
+	}
+
+	if len(args) >= 4 {
```
Comment (jsternberg) [inline query/compile.go]: This isn't possible. The portion above expects the number of arguments to be 3 or less.

File: query/compile.go [hunk 3]
```diff
@@ -550,6 +552,61 @@ func (c *compiledField) compileMovingAverage(args []influxql.Expr) error {
...
+		switch arg3 := args[3].(type) {
+		case *influxql.StringLiteral:
+			switch arg3.String() {
+			case "'exponential_moving_average'":
```
Comment (jsternberg) [inline query/compile.go]: The string here should not have single quotes. That will get removed on its own. I think using a name of `exponential_moving_average` isn't the best. Maybe this should just be `exponential`?

File: query/functions.go [hunk 4]
```diff
@@ -638,6 +639,59 @@ func (r *UnsignedMovingAverageReducer) Emit() []FloatPoint {
...
+	r.aggregate(float64(p.Value), p.Time)
+}
+func (r *ExponentialMovingAverageReducer) aggregate(v float64, t int64) {
+	r.v = r.ema.Add(v)
+	r.t = t
+	r.count++
+}
+func (r *ExponentialMovingAverageReducer) Emit() []FloatPoint {
+	if r.count <= r.holdPeriod {
+		return []FloatPoint(nil)
```
Comment (jsternberg) [inline query/functions.go]: I think you can just return `nil`.

File: query/select.go [hunk 5]
```diff
@@ -288,6 +288,28 @@ func (b *exprIteratorBuilder) buildCallIterator(ctx context.Context, expr *influ
...
+			}
+
+			nHold := -1
+			if len(expr.Args) >= 3 {
+				nHold = int(expr.Args[2].(*influxql.IntegerLiteral).Val)
+			}
+
+			var warmupType string
+			if len(expr.Args) >= 4 {
+				warmupType = expr.Args[3].(*influxql.StringLiteral).String()
```
Comment (jsternberg) [inline query/select.go]: There's no need to do this. Just do `expr.Args[3].(*influxql.StringLiteral).Val` and you won't have to strip the quotes.

File: query/compile.go [hunk 6]
```diff
@@ -569,6 +569,9 @@ func (c *compiledField) compileExponentialMovingAverage(name string, args []infl
 	if len(args) >= 3 {
 		switch arg2 := args[2].(type) {
 		case *influxql.IntegerLiteral:
+			if name == "triple_exponential_average" && arg2.Val < 1 && arg2.Val != -1 {
+				return fmt.Errorf("%s hold period must be greater than or equal to 1", name)
```
Comment (jsternberg) [inline query/compile.go]: Make some mention here about -1 being the option to set it to the default value. Maybe zero should set it to the default value instead? I'm not really sure I care which one signals the default value as long as one of them does and the error message mentions it.

File: query/select.go [hunk 7]
```diff
@@ -342,6 +342,28 @@ func (b *exprIteratorBuilder) buildCallIterator(ctx context.Context, expr *influ
...
```
Comment (jsternberg) [inline query/select.go]: Same things I mentioned before apply here too.

**Comment by jsternberg** [comment]
@phemmer above are my comments. If you have time, can you take a look at those and then you can push back to your own branch if you are OK with the changes I made?

File: query/compile.go (hunk 3 above)
Comment (phemmer) [reply query/compile.go]: Following from your other comments, I'm assuming you mean if `arg3.Val` is used instead. Valid point. I probably just didn't notice `Val` when I was doing the discovery & implementation.

As for the name, yes it's long, but it makes it clear what it is. "exponential" is ambiguous. Most other technical analysis applications just refer to it as "ema", but influxdb doesn't abbreviate anything, and the function name is also "exponential_moving_average", so I kept the long name for consistency.
Perhaps support both? `ema` and `exponential_moving_average`?

**Comment by jsternberg** [comment]
@natezimmer see #9260. This was just a PR for aiding in collaboration on this. #9260 is the real one. These functions should be available in the upcoming 1.6 release.

**Pull request closed without merging**